db/write_batch.cc

```diff
@@ -1597,6 +1602,22 @@ class MemTableInserter : public WriteBatch::Handler {
         flush_scheduler_->ScheduleFlush(cfd);
       }
     }
+    // check if memtable_list size exceeds max_write_buffer_size_to_maintain
+    // TODO: need to check if cfd have already been added to
+    // trim_history_scheduler_
+    if (trim_history_scheduler_ != nullptr) {
+      auto* cfd = cf_mems_->current();
+      assert(cfd != nullptr);
+      if (cfd->imm()->imm_trim_needed.load(std::memory_order_relaxed) ==
+              false &&
+          cfd->mem()->ApproximateMemoryUsage() +
```

Comment:
I checked the implementation of `ApproximateMemoryUsage()`. It's not a free call. It grabs the concurent arena mutex or something like that. We should try to call it twice for every write. We already called it once in cfd->mem()->ShouldScheduleFlush(). Can we reuse the value?

Reply:
@miasantreble I believe this is unresolved? Any plan to address this?

Reply:
Also, if you have decided to go with a separate `trim_history_scheduler`, is `imm_trim_needed` still needed?

Reply:
> We already called it once in cfd->mem()->ShouldScheduleFlush(). Can we reuse the value?
Do you mean `ShouldFlushNow()` because `ShouldScheduleFlush()` doesn't seem to call `ApproximateMemoryUsage()`. Maybe we could add a memtable variable to Memtable to keep track of the value of last call to `ApproximateMemoryUsage()` and read that value in `CheckMemtableFull()`?

Reply:
My second comment is not related to @siying's comment.
Since a cfd will be added to `TrimHistoryScheduler` which will be processed for history trimming during the preprocess phase of next write, I wonder whether you need `MemTableList::imm_trim_needed`.
If a cfd is in `trim_history_scheduler`, its `imm_trim_needed` must be true. And vice versa.
Does `imm_trim_needed` provide any extra information, or do you have something planned in mind?

Reply:
@riversand963 you are right, if we are using two queues, the boolean in memtable list is redundant. will remove it in the next upload

Reply:
@riversand963 I checked the code again, `imm_trim_needed` was added so that we don't repeatedly add a cfd to `FlushScheduler`. in other words, to keep this assert happy: https://github.com/facebook/rocksdb/blob/master/db/flush_scheduler.cc#L17